frontend: Add the clock-icon before the timestamp.

nikhilmaske-2001 opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) --> Add the clock-icon before the timestamp. Solves one of the part of issue #17293.

Related thread : https://chat.zulip.org/#narrow/stream/6-frontend/topic/time.20widget


**Testing plan:** <!-- How have you tested? -->


**GIFs or screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->
![clock-icon](https://user-images.githubusercontent.com/59444243/108059422-7cae0600-707b-11eb-9077-2d8447154c96.png)


<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

showell commented:
I think we typically put icons inside of an `<i>` tag, which you can see if you inspect the items in the gear menu.  I think this leads to a cleaner approach, as we can then set the spacing as margin, rather than prepending a space to the text.

timabbott commented:
Agreed.  Also I think `fa-clock-o` is the right icon, so well done there.

andersk commented on static/js/rendered_markdown.js:
```diff
@@ -159,7 +159,9 @@ exports.update_elements = (content) => {
-            $(this).text(rendered_time.text);
+            // fa fa-clock-o is the class for clock-icon
+            const rendered_foo = `<i class="fa fa-clock-o"></i>`;
+            $(this).html(rendered_foo + rendered_time.text);
```
Do not interpolate unescaped text into HTML.

showell commented on frontend_tests/node_tests/rendered_markdown.js:
```diff
@@ -181,7 +181,7 @@ run_test("timestamp", () => {
 
     // Final asserts
-    assert.equal($timestamp.text(), "Thu, Jan 1 1970, 12:00 AM");
+    assert.equal($timestamp.text(), "never-been-set");
```
You can't make tests pass by simply changing the expected values.  The `never-been-set` here should be a clue that you need to change your test setup here, or possibly change the implementation to be easier to test.

andersk commented on static/js/rendered_markdown.js:
```diff
@@ -159,7 +159,8 @@ exports.update_elements = (content) => {
             const text = $(this).text();
             const rendered_time = timerender.render_markdown_timestamp(timestamp, text);
-            $(this).text(rendered_time.text);
+            $(this).empty();
+            $(this).append(`<i class="fa fa-clock-o"></i>`, rendered_time.text);
```
`append` still interprets the string `rendered_time.text` as HTML.

```suggestion
            $(this).text(rendered_time.text)
            $(this).prepend($("<i>", {class: "fa fa-clock-o"}));
```

showell commented:
I think the consensus here is to use a template.  See https://chat.zulip.org/#narrow/stream/92-learning/topic/Do.20not.20interpolate.20unescaped.20text.20into.20HTML/near/1120147

nikhilmaske-2001 commented:
> I think the consensus here is to use a template.  See https://chat.zulip.org/#narrow/stream/92-learning/topic/Do.20not.20interpolate.20unescaped.20text.20into.20HTML/near/1120147

Yes I am working on it.

nikhilmaske-2001 commented on static/js/rendered_markdown.js:
```diff
@@ -161,8 +162,14 @@ export const update_elements = (content) => {
             const text = $(this).text();
             const rendered_time = timerender.render_markdown_timestamp(timestamp, text);
+            // Update the value of timestamp text to pass the tests.
             $(this).text(rendered_time.text);
```
The only reason to keep this line is, because it updates the value of timestamp text which is used [here](https://github.com/zulip/zulip/blob/master/frontend_tests/node_tests/rendered_markdown.js#L184) for testing.

timabbott replied:
We could probably just make the test check `.html()` instead, right?

nikhilmaske-2001 replied:
Sorry, I didn't understand your question, Can you elaborate your question a little more?

Signior-X replied:
@nikhilmaske-2001 He means that you can change the test accordingly to the new way you are using it. So, removing the older way in the actual code and updating the tests.

nikhilmaske-2001 replied:
Thanks @Signior-X for the explanation.

nikhilmaske-2001 commented:
As per the previous Tim's [review](https://github.com/zulip/zulip/pull/17318#discussion_r589195454), I have updated the PR accordingly. I have changed `.text()` method to `.html()` in the test file as we are now rendering a template.

timabbott commented:
Merged, after a few tweaks to the commit message and renaming the timeplate to `markdown_timestamp`, since `show_timestamp` is pretty ambiguous as to what part of the product it implements.  I also reordered the bit of logic to be a bit more readable / have less line-wrapping and not dodge our `.html()` linters.

Thanks for the contribution @nikhilmaske-2001!
